vcert/connection_tpp_common.py

```diff
@@ -0,0 +1,299 @@
+        raise NotImplementedError
+
+    # TODO: Need to add service generated CSR implementation
+    def request_cert(self, request, zone):
+        if not request.csr:
+            request.build_csr()
+        request_data = {"PolicyDN": self._get_policy_dn(zone),
+                        "PKCS10": request.csr,
+                        "ObjectName": request.friendly_name,
+                        "DisableAutomaticRenewal": "true"}
+        if request.origin:
+            request_data["Origin"] = request.origin
+            ca_origin = {"Name": "Origin", "Value": request.origin}
+            if request_data.get("CASpecificAttributes"):
+                request_data["CASpecificAttributes"].append(ca_origin)
+            else:
+                request_data["CASpecificAttributes"] = [ca_origin]
+        status, data = self._post(self.urls[URLS.CERTIFICATE_REQUESTS], data=request_data)
+        if status == HTTPStatus.OK:
+            request.id = data['CertificateDN']
+            log.debug("Certificate sucessfully requested with request id %s." % request.id)
+            return True
+        else:
+            log.error("Request status is not %s. %s." % HTTPStatus.OK, status)
+            raise CertificateRequestError
+
+    def retrieve_cert(self, certificate_request):
+        log.debug("Getting certificate status for id %s" % certificate_request.id)
+
+        retrieve_request = dict(CertificateDN=certificate_request.id, Format="base64", IncludeChain='true')
+
+        if certificate_request.chain_option == "last":
+            retrieve_request['RootFirstOrder'] = 'false'
+            retrieve_request['IncludeChain'] = 'true'
+        elif certificate_request.chain_option == "first":
+            retrieve_request['RootFirstOrder'] = 'true'
+            retrieve_request['IncludeChain'] = 'true'
+        elif certificate_request.chain_option == "ignore":
+            retrieve_request['IncludeChain'] = 'false'
+        else:
+            log.error("chain option %s is not valid" % certificate_request.chain_option)
+            raise ClientBadData
+
+        status, data = self._post(self.urls[URLS.CERTIFICATE_RETRIEVE], data=retrieve_request)
+        if status == HTTPStatus.OK:
+            pem64 = data['CertificateData']
+            pem = base64.b64decode(pem64)
+            return parse_pem(pem.decode(), certificate_request.chain_option)
+        elif status == HTTPStatus.ACCEPTED:
+            log.debug(data['Status'])
+            return None
+        else:
+            log.error("Status is not %s. %s" % HTTPStatus.OK, status)
+            raise ServerUnexptedBehavior
+
+    def revoke_cert(self, request):
+        if not (request.id or request.thumbprint):
+            raise ClientBadData
+        d = {
+            "Disable": request.disable
+        }
+        if request.reason:
+            d["Reason"] = request.reason
+        if request.id:
+            d["CertificateDN"] = request.id
+        elif request.thumbprint:
```

Comment:
Is id and thumbprint either or, as I see if 'id' else 'thumbprint' ??

Reply:
They are not mutually exclusive since the API method can receive both values in the request. However, it is expected to receive one or the other as identifier of the certificate. This code has not been modified in this PR, just moved to a new file/resource.